Set default file extension in Windows save dialog

samangh opened the pull request:
First of all, thank you for your work on portable-file-dialogs. It's quite useful :-).
 
This pull request will set the default file extension when using the file save dialog in Windows.  The default file extension is set to the first file extension provided in the filter pattern.

It turns out that the `GetSaveFileNameW` function in Windows does not automatically get the file extension from the filters, instead the the default extension must be manually set.  So this is what this pull request does.

This should close issue #43.

gamagan commented:
Questions:
1. What happens if the user selects a different file extension from the dropdown? Does the file 
a) Still get saved with the default extension
b) Uses the new extension selected
c) Uses no extension
d) Other

2) Seem to me like regular expressions are overkill for getting the extension value. Can't a few calls to std::string::find() accomplish that?

samangh committed:
Include default file extension in Windows save dialog

With this commit, in a Windows file extension, if the user has not
entered a file extension, then the file extension from the filter
pattern will be used.

samangh commented:
@gamagan, thanks for the questions. They caused me to look at this more closely and do some experimentation.

I was actually incorrect in my previous message. As long as the `lpstrDefExt` pointer is allocated, Windows does the right thing and uses the extension defined in the pattern filter. The actual content of `lpstrDefExt` is only used if  the filter pattern is blank or is `*.*` and the user has not entered an extension.

I have updated my pull request to match this. In the new pull request, things should work as expected. So, for example let's say you do:

```cpp
pfd::save_file("Select a file", ".",                            
              { "Image Files", "*.png *.jpg *.jpeg *.bmp",
                "AudioFiles", "*.wav *.mp3",
                "All Files", "*" });
```

The default extension will be `.png` if `Image Files` is selected, `.wav` if `Audio Files` is selected, and no extension is added if `All files` is selected.

samangh commented:
@samhocevar, I was wondering if you've had a chance to look at the PR? I think it's a nice uncontroversial 2-line change.

JVital2013 committed:
[macOS] Fix save dialog when absolute path is included

samangh commented:
@samhocevar? Any chance of merging this pull request?

tgalaj commented:
> @samhocevar? Any chance of merging this pull request?

Yup, accepting this PR would make our lives much, much easier @samhocevar  

This small library is really great, it would be a waste to abandon it.

christophe-f8 commented:
I have the same problem, the extension is not automatically added and I'd like this to be merged too. For now, I'm using @samangh branch.

I did further improve this though so I though I would share. I didn't wanted to open yet another PR. @samangh maybe you can update yours ?
Anyway, here are the two small modifications:
- increase the extension length to 64 characters (with only 3 characters it wouldn't work with ".json" for example).
- Solved a similar bug in file open dialogs: if the user types a name of a file that does not exists, there was no popup showing (works well including with multiple file selections).

In the end, the few lines at that place in the file become:
`       if (in_type == type::save)
       {
           auto wextension = std::wstring(64, L'\0');
           ofn.lpstrDefExt = (LPWSTR)wextension.data();
           
           if (!(options & opt::force_overwrite))
               ofn.Flags |= OFN_OVERWRITEPROMPT;

           dll::proc<BOOL WINAPI (LPOPENFILENAMEW)> get_save_file_name(comdlg32, "GetSaveFileNameW");
           if (get_save_file_name(&ofn) == 0)
               return "";
           return internal::wstr2str(woutput.c_str());
       }
       else
       {
           if (options & opt::multiselect)
               ofn.Flags |= OFN_ALLOWMULTISELECT;
           ofn.Flags |= OFN_PATHMUSTEXIST;
           if (in_type == type::open)
               ofn.Flags |= OFN_FILEMUSTEXIST;

           dll::proc<BOOL WINAPI (LPOPENFILENAMEW)> get_open_file_name(comdlg32, "GetOpenFileNameW");
           if (get_open_file_name(&ofn) == 0)
               return "";
       }`

samangh committed:
Don't set minimum CMake version

The required version number (3.5) us so old that CMake is now
complaining about this.

Removing a required CMake version fixes this. We don't use anything
complicated so any CMake version will be compatible.

samangh commented:
@christophe-f8 Sure, I can look at merging this in my branch.

As this is meant to be a portable library I'll have to check that the same behaviour  is also implemented in other environments.

samangh committed:
Add `BUILD_EXAMPLES` option to CMake

samangh committed:
Set extenion length to 64 chars in Window file save dialogs

samangh committed:
Show error if file does not exist in Windows open file dialog

Suggested in https://github.com/samhocevar/portable-file-dialogs/pull/86.

samangh commented:
@christophe-f8, your suggested changes are now in my branch.

samangh committed:
Merge pull request #1 from JVital2013/main

[macOS] Fix save dialog when absolute path is included

samangh committed:
windows: hide read-only option from open file dialog

From samhocevar/portable-file-dialogs#83

samangh committed:
kde: fix detction of cancel button

Fixes samhocevar/portable-file-dialogs#92

samangh committed:
windows: ensure dialogs are modal

`GetActiveWindow()` will return NULL if called asynchronously, as it's
calling thread will not have a message loop. To work around this, use
`GetForegroundWindow()` if `GetActiveWindow()` is null.

This will cause any dialogs to be modal.

Fixes samhocevar/portable-file-dialogs#82 and
samhocevar/portable-file-dialogs#93

samangh committed:
kde:: suport `abort_retry_ignore`

samangh committed:
Merge remote-tracking branch 'upstream/main'
